File: fud-ai/android/app/src/main/java/com/apoorvdarshan/calorietracker/export/DiaryExporter.kt
package com.apoorvdarshan.calorietracker.export

import com.apoorvdarshan.calorietracker.R

import com.apoorvdarshan.calorietracker.models.FoodEntry
import com.apoorvdarshan.calorietracker.models.FoodSource
import com.apoorvdarshan.calorietracker.models.MealType
import com.apoorvdarshan.calorietracker.models.UserProfile
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import kotlin.math.max
import kotlin.math.roundToInt

enum class DiaryFormat(val label: String, val ext: String, val mime: String) {
    JSON("JSON", "json", "application/json"),
    MARKDOWN("Markdown", "md", "text/markdown"),
    CSV("CSV", "csv", "text/csv"),
}

enum class DiaryRange(val label: String, val labelRes: Int) {
    TODAY("Today", R.string.export_range_today),
    THIS_WEEK("This week", R.string.export_range_week),
    THIS_MONTH("This month", R.string.export_range_month),
    ALL_TIME("All time", R.string.export_range_all),
    CUSTOM("Custom", R.string.export_range_custom),
}

/**
 * Builds a shareable food-diary file (JSON / Markdown / CSV) from the local log.
 * Pure logic — the caller supplies the entries, the current profile, and a meal
 * display-name resolver (so string resources stay in the UI layer).
 */
object DiaryExporter {

    private val zone: ZoneId get() = ZoneId.systemDefault()
    private val dayFmt: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
    private val timeFmt: DateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm")

    private fun LocalDate.of(entry: FoodEntry): Boolean =
        entry.timestamp.atZone(zone).toLocalDate() == this

    fun resolveRange(
        range: DiaryRange,
        customStart: LocalDate,
        customEnd: LocalDate,
        entries: List<FoodEntry>,
    ): Pair<LocalDate, LocalDate> {
        val today = LocalDate.now()
        return when (range) {
            DiaryRange.TODAY -> today to today
            DiaryRange.THIS_WEEK -> today.with(DayOfWeek.MONDAY) to today
            DiaryRange.THIS_MONTH -> today.withDayOfMonth(1) to today
            DiaryRange.ALL_TIME -> {
                val earliest = entries.minOfOrNull { it.timestamp.atZone(zone).toLocalDate() } ?: today
                earliest to today
            }
            DiaryRange.CUSTOM -> customStart to customEnd
        }
    }

    /** Returns (filename, content) or null if nothing is logged in the range. */
    fun build(
        entries: List<FoodEntry>,
        start: LocalDate,
        end: LocalDate,
        format: DiaryFormat,
        profile: UserProfile?,
        mealDisplay: (MealType) -> String,
    ): Pair<String, String>? {
        val lo = if (start.isAfter(end)) end else start
        val hi = if (start.isAfter(end)) start else end

        val byDay: Map<LocalDate, List<FoodEntry>> = entries
            .filter {
                val d = it.timestamp.atZone(zone).toLocalDate()
                !d.isBefore(lo) && !d.isAfter(hi)
            }
            .groupBy { it.timestamp.atZone(zone).toLocalDate() }
            .toSortedMap()
        if (byDay.isEmpty()) return null

        val targets = Targets(
            calories = profile?.effectiveCalories ?: 0,
            protein = (profile?.effectiveProtein ?: 0).toDouble(),
            carbs = (profile?.effectiveCarbs ?: 0).toDouble(),
            fat = (profile?.effectiveFat ?: 0).toDouble(),
        )

        val content = when (format) {
            DiaryFormat.JSON -> json(byDay, lo, hi, targets)
            DiaryFormat.MARKDOWN -> markdown(byDay, lo, hi, targets, mealDisplay)
            DiaryFormat.CSV -> csv(byDay)
        }
        val name = "Fud-Food-Diary-${dayFmt.format(lo)}_to_${dayFmt.format(hi)}.${format.ext}"
        return name to content
    }

    // --- internal helpers ---

    private data class Targets(val calories: Int, val protein: Double, val carbs: Double, val fat: Double)

    /** A day's entries grouped by meal in enum order, each sorted by time ascending. */
    private fun meals(dayEntries: List<FoodEntry>): List<Pair<MealType, List<FoodEntry>>> =
        MealType.values().mapNotNull { mt ->
            val items = dayEntries.filter { it.mealType == mt }.sortedBy { it.timestamp }
            if (items.isEmpty()) null else mt to items
        }

    private fun totals(dayEntries: List<FoodEntry>): DoubleArray {
        var cal = 0.0; var p = 0.0; var c = 0.0; var f = 0.0
        for (e in dayEntries) { cal += e.calories; p += e.protein; c += e.carbs; f += e.fat }
        return doubleArrayOf(cal, p, c, f)
    }

    fun sourceLabel(source: FoodSource): String =
        if (source == FoodSource.MANUAL) "manually_edited" else "ai_estimated"

    private fun r1(x: Double): Double = (x * 10).roundToInt() / 10.0
    private fun optionalNumber(x: Double?, missing: String = ""): String = x?.let { r1(it).toString() } ?: missing
    private fun time(entry: FoodEntry): String = timeFmt.format(entry.timestamp.atZone(zone))

    // --- JSON ---

    @Serializable private data class Macro(val calories: Int, val protein_g: Double, val carbs_g: Double, val fat_g: Double)
    @Serializable private data class ItemDto(
        val name: String, val quantity_g: Double? = null, val calories: Int,
        val protein_g: Double, val carbs_g: Double, val fat_g: Double,
        val sugar_g: Double? = null, val added_sugar_g: Double? = null, val fiber_g: Double? = null,
        val saturated_fat_g: Double? = null, val monounsaturated_fat_g: Double? = null,
        val polyunsaturated_fat_g: Double? = null, val cholesterol_mg: Double? = null,
        val sodium_mg: Double? = null, val potassium_mg: Double? = null, val trans_fat_g: Double? = null,
        val calcium_mg: Double? = null, val iron_mg: Double? = null, val magnesium_mg: Double? = null,
        val zinc_mg: Double? = null, val vitamin_a_mcg: Double? = null, val vitamin_c_mg: Double? = null,
        val vitamin_d_mcg: Double? = null, val vitamin_b12_mcg: Double? = null,
        val vitamin_e_mg: Double? = null, val vitamin_k_mcg: Double? = null,
        val folate_mcg: Double? = null, val omega3_g: Double? = null,
        val time: String, val source: String, val note: String? = null,
    )
    @Serializable private data class MealDto(val type: String, val items: List<ItemDto>)
    @Serializable private data class DayDto(val date: String, val totals: Macro, val targets: Macro, val remaining: Macro, val meals: List<MealDto>)
    @Serializable private data class RangeDto(val start: String, val end: String)
    @Serializable private data class MetaDto(val app: String, val format_version: String, val date_range: RangeDto)
    @Serializable private data class Doc(val export: MetaDto, val days: List<DayDto>)

    private val jsonPretty = Json { prettyPrint = true; encodeDefaults = true }

    private fun json(byDay: Map<LocalDate, List<FoodEntry>>, lo: LocalDate, hi: LocalDate, t: Targets): String {
        val days = byDay.map { (date, dayEntries) ->
            val tot = totals(dayEntries)
            val mealDtos = meals(dayEntries).map { (mt, items) ->
                MealDto(
                    type = mt.name.lowercase(),
                    items = items.map { e ->
                        ItemDto(
                            name = e.name,
                            quantity_g = e.servingSizeGrams?.let { r1(it) },
                            calories = e.calories,
                            protein_g = r1(e.protein), carbs_g = r1(e.carbs), fat_g = r1(e.fat),
                            sugar_g = e.sugar?.let { r1(it) }, added_sugar_g = e.addedSugar?.let { r1(it) },
                            fiber_g = e.fiber?.let { r1(it) }, saturated_fat_g = e.saturatedFat?.let { r1(it) },
                            monounsaturated_fat_g = e.monounsaturatedFat?.let { r1(it) },
                            polyunsaturated_fat_g = e.polyunsaturatedFat?.let { r1(it) },
                            cholesterol_mg = e.cholesterol?.let { r1(it) }, sodium_mg = e.sodium?.let { r1(it) },
                            potassium_mg = e.potassium?.let { r1(it) }, trans_fat_g = e.transFat?.let { r1(it) },
                            calcium_mg = e.calcium?.let { r1(it) }, iron_mg = e.iron?.let { r1(it) },
                            magnesium_mg = e.magnesium?.let { r1(it) }, zinc_mg = e.zinc?.let { r1(it) },
                            vitamin_a_mcg = e.vitaminA?.let { r1(it) }, vitamin_c_mg = e.vitaminC?.let { r1(it) },
                            vitamin_d_mcg = e.vitaminD?.let { r1(it) }, vitamin_b12_mcg = e.vitaminB12?.let { r1(it) },
                            vitamin_e_mg = e.vitaminE?.let { r1(it) }, vitamin_k_mcg = e.vitaminK?.let { r1(it) },
                            folate_mcg = e.folate?.let { r1(it) }, omega3_g = e.omega3?.let { r1(it) },
                            time = time(e), source = sourceLabel(e.source),
                            note = e.customNote?.takeIf { it.isNotBlank() },
                        )
                    },
                )
            }
            DayDto(
                date = dayFmt.format(date),
                totals = Macro(tot[0].roundToInt(), r1(tot[1]), r1(tot[2]), r1(tot[3])),
                targets = Macro(t.calories, t.protein, t.carbs, t.fat),
                remaining = Macro(
                    max(0, t.calories - tot[0].roundToInt()),
                    r1(max(0.0, t.protein - tot[1])),
                    r1(max(0.0, t.carbs - tot[2])),
                    r1(max(0.0, t.fat - tot[3])),
                ),
                meals = mealDtos,
            )
        }
        val doc = Doc(
            export = MetaDto("Fud AI", "1.1", RangeDto(dayFmt.format(lo), dayFmt.format(hi))),
            days = days,
        )
        return jsonPretty.encodeToString(Doc.serializer(), doc)
    }

    // --- Markdown ---

    private fun markdown(
        byDay: Map<LocalDate, List<FoodEntry>>,
        lo: LocalDate, hi: LocalDate, t: Targets,
        mealDisplay: (MealType) -> String,
    ): String {
        val sb = StringBuilder()
        sb.append("# Food diary export\n")
        sb.append("Date range: ${dayFmt.format(lo)} to ${dayFmt.format(hi)}\n")
        sb.append("Generated by Fud AI\n")
        for ((date, dayEntries) in byDay) {
            val tot = totals(dayEntries)
            sb.append("\n## ${dayFmt.format(date)}\n")
            sb.append("Totals:\n")
            sb.append("- Calories: ${tot[0].roundToInt()} / ${t.calories} kcal\n")
            sb.append("- Protein: ${r1(tot[1])} / ${t.protein.roundToInt()} g\n")
            sb.append("- Carbs: ${r1(tot[2])} / ${t.carbs.roundToInt()} g\n")
            sb.append("- Fat: ${r1(tot[3])} / ${t.fat.roundToInt()} g\n")
            for ((mt, items) in meals(dayEntries)) {
                sb.append("### ${mealDisplay(mt)}\n")
                sb.append("| Time | Food | Weight | Calories | Protein (g) | Carbs (g) | Fat (g) | Sugar (g) | Added sugar (g) | Fiber (g) | Saturated fat (g) | Monounsaturated fat (g) | Polyunsaturated fat (g) | Cholesterol (mg) | Sodium (mg) | Potassium (mg) | Trans fat (g) | Calcium (mg) | Iron (mg) | Magnesium (mg) | Zinc (mg) | Vitamin A (mcg) | Vitamin C (mg) | Vitamin D (mcg) | Vitamin B12 (mcg) | Vitamin E (mg) | Vitamin K (mcg) | Folate (mcg) | Omega-3 (g) | Source |\n")
                sb.append("|---|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---|\n")
                for (e in items) {
                    val weight = e.servingSizeGrams?.let { "${it.roundToInt()} g" } ?: "-"
                    val food = e.name.replace("|", "/")
                    val cells = listOf(
                        time(e), food, weight, e.calories.toString(),
                        r1(e.protein).toString(), r1(e.carbs).toString(), r1(e.fat).toString(),
                        optionalNumber(e.sugar, "-"), optionalNumber(e.addedSugar, "-"), optionalNumber(e.fiber, "-"),
                        optionalNumber(e.saturatedFat, "-"), optionalNumber(e.monounsaturatedFat, "-"),
                        optionalNumber(e.polyunsaturatedFat, "-"), optionalNumber(e.cholesterol, "-"),
                        optionalNumber(e.sodium, "-"), optionalNumber(e.potassium, "-"), optionalNumber(e.transFat, "-"),
                        optionalNumber(e.calcium, "-"), optionalNumber(e.iron, "-"), optionalNumber(e.magnesium, "-"),
                        optionalNumber(e.zinc, "-"), optionalNumber(e.vitaminA, "-"), optionalNumber(e.vitaminC, "-"),
                        optionalNumber(e.vitaminD, "-"), optionalNumber(e.vitaminB12, "-"), optionalNumber(e.vitaminE, "-"),
                        optionalNumber(e.vitaminK, "-"), optionalNumber(e.folate, "-"), optionalNumber(e.omega3, "-"),
                        sourceLabel(e.source),
                    )
                    sb.append("| ").append(cells.joinToString(" | ")).append(" |\n")
                }
            }
        }
        return sb.toString()
    }

    // --- CSV ---

    private fun csv(byDay: Map<LocalDate, List<FoodEntry>>): String {
        val sb = StringBuilder()
        sb.append("date,meal,time,food,weight_g,calories,protein_g,carbs_g,fat_g,sugar_g,added_sugar_g,fiber_g,saturated_fat_g,monounsaturated_fat_g,polyunsaturated_fat_g,cholesterol_mg,sodium_mg,potassium_mg,trans_fat_g,calcium_mg,iron_mg,magnesium_mg,zinc_mg,vitamin_a_mcg,vitamin_c_mg,vitamin_d_mcg,vitamin_b12_mcg,vitamin_e_mg,vitamin_k_mcg,folate_mcg,omega3_g,source,note\n")
        for ((date, dayEntries) in byDay) {
            val d = dayFmt.format(date)
            for ((mt, items) in meals(dayEntries)) {
                for (e in items) {
                    val cols = listOf(
                        d, mt.name.lowercase(), time(e), e.name,
                        e.servingSizeGrams?.roundToInt()?.toString() ?: "",
                        e.calories.toString(), r1(e.protein).toString(), r1(e.carbs).toString(), r1(e.fat).toString(),
                        optionalNumber(e.sugar), optionalNumber(e.addedSugar), optionalNumber(e.fiber),
                        optionalNumber(e.saturatedFat), optionalNumber(e.monounsaturatedFat),
                        optionalNumber(e.polyunsaturatedFat), optionalNumber(e.cholesterol), optionalNumber(e.sodium),
                        optionalNumber(e.potassium), optionalNumber(e.transFat), optionalNumber(e.calcium),
                        optionalNumber(e.iron), optionalNumber(e.magnesium), optionalNumber(e.zinc),
                        optionalNumber(e.vitaminA), optionalNumber(e.vitaminC), optionalNumber(e.vitaminD),
                        optionalNumber(e.vitaminB12), optionalNumber(e.vitaminE), optionalNumber(e.vitaminK),
                        optionalNumber(e.folate), optionalNumber(e.omega3),
                        sourceLabel(e.source), e.customNote ?: "",
                    )
                    sb.append(cols.joinToString(",") { csvEscape(it) }).append("\n")
                }
            }
        }
        return sb.toString()
    }

    private fun csvEscape(field: String): String =
        if (field.contains(',') || field.contains('"') || field.contains('\n')) {
            "\"" + field.replace("\"", "\"\"") + "\""
        } else field
}

File: fud-ai/ios/calorietracker/Services/DiaryExporter.swift
import Foundation

/// Output format for a food-diary export.
enum DiaryExportFormat: String, CaseIterable, Identifiable {
    case json, markdown, csv
    var id: String { rawValue }
    var label: String {
        switch self {
        case .json: return "JSON"
        case .markdown: return "Markdown"
        case .csv: return "CSV"
        }
    }
    var fileExtension: String {
        switch self {
        case .json: return "json"
        case .markdown: return "md"
        case .csv: return "csv"
        }
    }
}

/// Date-range presets for an export.
enum DiaryExportRange: String, CaseIterable, Identifiable {
    case today, thisWeek, thisMonth, allTime, custom
    var id: String { rawValue }
    var label: String {
        switch self {
        case .today: return "Today"
        case .thisWeek: return "This week"
        case .thisMonth: return "This month"
        case .allTime: return "All time"
        case .custom: return "Custom"
        }
    }
}

/// Builds a shareable food-diary file (JSON / Markdown / CSV) from the local log.
/// Pure logic — no UI. Everything comes from the already-loaded FoodStore + profile.
enum DiaryExporter {

    // MARK: - Public entry point

    /// Returns (filename, data) for the given inclusive day range, or nil if there is
    /// nothing logged in that range.
    static func build(
        from startDay: Date,
        to endDay: Date,
        format: DiaryExportFormat,
        foodStore: FoodStore,
        profile: UserProfile
    ) -> (filename: String, data: Data)? {
        build(
            from: startDay,
            to: endDay,
            format: format,
            entries: foodStore.entries,
            profile: profile
        )
    }

    /// Pure entry-based overload used by the app exporter and unit tests.
    static func build(
        from startDay: Date,
        to endDay: Date,
        format: DiaryExportFormat,
        entries: [FoodEntry],
        profile: UserProfile
    ) -> (filename: String, data: Data)? {
        let cal = Calendar.current
        let start = cal.startOfDay(for: min(startDay, endDay))
        let end = cal.startOfDay(for: max(startDay, endDay))

        var days: [DayBundle] = []
        var day = start
        while day <= end {
            let dayEntries = entries
                .filter { cal.isDate($0.timestamp, inSameDayAs: day) }
                .sorted { $0.timestamp > $1.timestamp }
            if !dayEntries.isEmpty {
                let groups = MealType.allCases.compactMap { meal -> FoodLogMealGroup? in
                    let mealEntries = dayEntries.filter { $0.mealType == meal }
                    guard !mealEntries.isEmpty else { return nil }
                    return FoodLogMealGroup(id: "export-\(meal.rawValue)", meal: meal, entries: mealEntries)
                }
                days.append(DayBundle(date: day, groups: groups))
            }
            guard let next = cal.date(byAdding: .day, value: 1, to: day) else { break }
            day = next
        }
        guard !days.isEmpty else { return nil }

        let targets = Targets(
            calories: profile.effectiveCalories,
            protein: Double(profile.effectiveProtein),
            carbs: Double(profile.effectiveCarbs),
            fat: Double(profile.effectiveFat)
        )

        let text: String
        switch format {
        case .json: text = json(days, start: start, end: end, targets: targets)
        case .markdown: text = markdown(days, start: start, end: end, targets: targets)
        case .csv: text = csv(days)
        }

        let name = "Fud-Food-Diary-\(dayFmt.string(from: start))_to_\(dayFmt.string(from: end)).\(format.fileExtension)"
        return (name, Data(text.utf8))
    }

    // MARK: - Range resolution

    /// Resolves a preset (or custom bounds) into an inclusive (startDay, endDay).
    static func resolve(_ range: DiaryExportRange, customStart: Date, customEnd: Date, foodStore: FoodStore) -> (Date, Date) {
        let cal = Calendar.current
        let today = cal.startOfDay(for: .now)
        switch range {
        case .today:
            return (today, today)
        case .thisWeek:
            let start = cal.dateInterval(of: .weekOfYear, for: today)?.start ?? today
            return (start, today)
        case .thisMonth:
            let start = cal.dateInterval(of: .month, for: today)?.start ?? today
            return (start, today)
        case .allTime:
            let earliest = foodStore.entries.map(\.timestamp).min().map { cal.startOfDay(for: $0) } ?? today
            return (earliest, today)
        case .custom:
            return (cal.startOfDay(for: customStart), cal.startOfDay(for: customEnd))
        }
    }

    // MARK: - Internal shapes

    private struct DayBundle { let date: Date; let groups: [FoodLogMealGroup] }
    private struct Targets { let calories: Int; let protein: Double; let carbs: Double; let fat: Double }

    private static func totals(_ groups: [FoodLogMealGroup]) -> (cal: Int, p: Double, c: Double, f: Double) {
        var cal = 0; var p = 0.0; var c = 0.0; var f = 0.0
        for g in groups { for e in g.entries { cal += e.calories; p += e.protein; c += e.carbs; f += e.fat } }
        return (cal, p, c, f)
    }

    static func sourceLabel(_ source: FoodSource) -> String {
        source == .manual ? "manually_edited" : "ai_estimated"
    }

    private nonisolated static func r1(_ x: Double) -> Double { (x * 10).rounded() / 10 }

    private nonisolated static func optionalNumber(_ x: Double?, missing: String = "") -> String {
        x.map { String(r1($0)) } ?? missing
    }

    private static let dayFmt: DateFormatter = {
        let f = DateFormatter(); f.locale = Locale(identifier: "en_US_POSIX"); f.dateFormat = "yyyy-MM-dd"; return f
    }()
    private static let timeFmt: DateFormatter = {
        let f = DateFormatter(); f.locale = Locale(identifier: "en_US_POSIX"); f.dateFormat = "HH:mm"; return f
    }()

    // MARK: - JSON

    private static func json(_ days: [DayBundle], start: Date, end: Date, targets: Targets) -> String {
        struct Macro: Encodable { let calories: Int; let protein_g: Double; let carbs_g: Double; let fat_g: Double }
        struct Item: Encodable {
            let name: String; let quantity_g: Double?; let calories: Int
            let protein_g: Double; let carbs_g: Double; let fat_g: Double
            let sugar_g: Double?; let added_sugar_g: Double?; let fiber_g: Double?
            let saturated_fat_g: Double?; let monounsaturated_fat_g: Double?; let polyunsaturated_fat_g: Double?
            let cholesterol_mg: Double?; let sodium_mg: Double?; let potassium_mg: Double?
            let trans_fat_g: Double?; let calcium_mg: Double?; let iron_mg: Double?
            let magnesium_mg: Double?; let zinc_mg: Double?
            let vitamin_a_mcg: Double?; let vitamin_c_mg: Double?; let vitamin_d_mcg: Double?
            let vitamin_b12_mcg: Double?; let vitamin_e_mg: Double?; let vitamin_k_mcg: Double?
            let folate_mcg: Double?; let omega3_g: Double?
            let time: String; let source: String; let note: String?
        }
        struct Meal: Encodable { let type: String; let items: [Item] }
        struct Day: Encodable { let date: String; let totals: Macro; let targets: Macro; let remaining: Macro; let meals: [Meal] }
        struct Meta: Encodable { struct Range: Encodable { let start: String; let end: String }
            let app: String; let format_version: String; let date_range: Range }
        struct Doc: Encodable { let export: Meta; let days: [Day] }

        let dayDocs: [Day] = days.map { bundle in
            let t = totals(bundle.groups)
            let meals: [Meal] = bundle.groups.map { g in
                Meal(type: g.meal.rawValue, items: g.entries.map { e in
                    Item(name: e.name, quantity_g: e.servingSizeGrams.map(r1), calories: e.calories,
                         protein_g: r1(e.protein), carbs_g: r1(e.carbs), fat_g: r1(e.fat),
                         sugar_g: e.sugar.map(r1), added_sugar_g: e.addedSugar.map(r1), fiber_g: e.fiber.map(r1),
                         saturated_fat_g: e.saturatedFat.map(r1),
                         monounsaturated_fat_g: e.monounsaturatedFat.map(r1),
                         polyunsaturated_fat_g: e.polyunsaturatedFat.map(r1),
                         cholesterol_mg: e.cholesterol.map(r1), sodium_mg: e.sodium.map(r1),
                         potassium_mg: e.potassium.map(r1), trans_fat_g: e.transFat.map(r1),
                         calcium_mg: e.calcium.map(r1), iron_mg: e.iron.map(r1),
                         magnesium_mg: e.magnesium.map(r1), zinc_mg: e.zinc.map(r1),
                         vitamin_a_mcg: e.vitaminA.map(r1), vitamin_c_mg: e.vitaminC.map(r1),
                         vitamin_d_mcg: e.vitaminD.map(r1), vitamin_b12_mcg: e.vitaminB12.map(r1),
                         vitamin_e_mg: e.vitaminE.map(r1), vitamin_k_mcg: e.vitaminK.map(r1),
                         folate_mcg: e.folate.map(r1), omega3_g: e.omega3.map(r1),
                         time: timeFmt.string(from: e.timestamp), source: sourceLabel(e.source),
                         note: (e.customNote?.isEmpty == false) ? e.customNote : nil)
                })
            }
            return Day(
                date: dayFmt.string(from: bundle.date),
                totals: Macro(calories: t.cal, protein_g: r1(t.p), carbs_g: r1(t.c), fat_g: r1(t.f)),
                targets: Macro(calories: targets.calories, protein_g: targets.protein, carbs_g: targets.carbs, fat_g: targets.fat),
                remaining: Macro(calories: max(0, targets.calories - t.cal),
                                 protein_g: r1(max(0, targets.protein - t.p)),
                                 carbs_g: r1(max(0, targets.carbs - t.c)),
                                 fat_g: r1(max(0, targets.fat - t.f))),
                meals: meals
            )
        }
        let doc = Doc(export: Meta(app: "Fud AI", format_version: "1.1",
                                   date_range: Meta.Range(start: dayFmt.string(from: start), end: dayFmt.string(from: end))),
                      days: dayDocs)
        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .withoutEscapingSlashes]
        return (try? enc.encode(doc)).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
    }

    // MARK: - Markdown

    private static func markdown(_ days: [DayBundle], start: Date, end: Date, targets: Targets) -> String {
        var s = "# Food diary export\n"
        s += "Date range: \(dayFmt.string(from: start)) to \(dayFmt.string(from: end))\n"
        s += "Generated by Fud AI\n"
        for bundle in days {
            let t = totals(bundle.groups)
            s += "\n## \(dayFmt.string(from: bundle.date))\n"
            s += "Totals:\n"
            s += "- Calories: \(t.cal) / \(targets.calories) kcal\n"
            s += "- Protein: \(r1(t.p)) / \(Int(targets.protein)) g\n"
            s += "- Carbs: \(r1(t.c)) / \(Int(targets.carbs)) g\n"
            s += "- Fat: \(r1(t.f)) / \(Int(targets.fat)) g\n"
            for g in bundle.groups {
                s += "### \(g.meal.displayName)\n"
                s += "| Time | Food | Weight | Calories | Protein (g) | Carbs (g) | Fat (g) | Sugar (g) | Added sugar (g) | Fiber (g) | Saturated fat (g) | Monounsaturated fat (g) | Polyunsaturated fat (g) | Cholesterol (mg) | Sodium (mg) | Potassium (mg) | Trans fat (g) | Calcium (mg) | Iron (mg) | Magnesium (mg) | Zinc (mg) | Vitamin A (mcg) | Vitamin C (mg) | Vitamin D (mcg) | Vitamin B12 (mcg) | Vitamin E (mg) | Vitamin K (mcg) | Folate (mcg) | Omega-3 (g) | Source |\n"
                s += "|---|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---|\n"
                for e in g.entries {
                    let weight = e.servingSizeGrams.map { "\(Int($0)) g" } ?? "-"
                    let food = e.name.replacingOccurrences(of: "|", with: "/")
                    let cells: [String] = [
                        timeFmt.string(from: e.timestamp), food, weight,
                        String(e.calories), String(r1(e.protein)), String(r1(e.carbs)), String(r1(e.fat)),
                        optionalNumber(e.sugar, missing: "-"), optionalNumber(e.addedSugar, missing: "-"),
                        optionalNumber(e.fiber, missing: "-"), optionalNumber(e.saturatedFat, missing: "-"),
                        optionalNumber(e.monounsaturatedFat, missing: "-"), optionalNumber(e.polyunsaturatedFat, missing: "-"),
                        optionalNumber(e.cholesterol, missing: "-"), optionalNumber(e.sodium, missing: "-"),
                        optionalNumber(e.potassium, missing: "-"), optionalNumber(e.transFat, missing: "-"),
                        optionalNumber(e.calcium, missing: "-"), optionalNumber(e.iron, missing: "-"),
                        optionalNumber(e.magnesium, missing: "-"), optionalNumber(e.zinc, missing: "-"),
                        optionalNumber(e.vitaminA, missing: "-"), optionalNumber(e.vitaminC, missing: "-"),
                        optionalNumber(e.vitaminD, missing: "-"), optionalNumber(e.vitaminB12, missing: "-"),
                        optionalNumber(e.vitaminE, missing: "-"), optionalNumber(e.vitaminK, missing: "-"),
                        optionalNumber(e.folate, missing: "-"), optionalNumber(e.omega3, missing: "-"),
                        sourceLabel(e.source)
                    ]
                    s += "| " + cells.joined(separator: " | ") + " |\n"
                }
            }
        }
        return s
    }

    // MARK: - CSV

    private static func csv(_ days: [DayBundle]) -> String {
        var s = "date,meal,time,food,weight_g,calories,protein_g,carbs_g,fat_g,sugar_g,added_sugar_g,fiber_g,saturated_fat_g,monounsaturated_fat_g,polyunsaturated_fat_g,cholesterol_mg,sodium_mg,potassium_mg,trans_fat_g,calcium_mg,iron_mg,magnesium_mg,zinc_mg,vitamin_a_mcg,vitamin_c_mg,vitamin_d_mcg,vitamin_b12_mcg,vitamin_e_mg,vitamin_k_mcg,folate_mcg,omega3_g,source,note\n"
        for bundle in days {
            let date = dayFmt.string(from: bundle.date)
            for g in bundle.groups {
                for e in g.entries {
                    let cols: [String] = [
                        date, g.meal.rawValue, timeFmt.string(from: e.timestamp), e.name,
                        e.servingSizeGrams.map { String(Int($0)) } ?? "",
                        String(e.calories), String(r1(e.protein)), String(r1(e.carbs)), String(r1(e.fat)),
                        optionalNumber(e.sugar), optionalNumber(e.addedSugar), optionalNumber(e.fiber),
                        optionalNumber(e.saturatedFat), optionalNumber(e.monounsaturatedFat),
                        optionalNumber(e.polyunsaturatedFat), optionalNumber(e.cholesterol),
                        optionalNumber(e.sodium), optionalNumber(e.potassium), optionalNumber(e.transFat),
                        optionalNumber(e.calcium), optionalNumber(e.iron), optionalNumber(e.magnesium),
                        optionalNumber(e.zinc), optionalNumber(e.vitaminA), optionalNumber(e.vitaminC),
                        optionalNumber(e.vitaminD), optionalNumber(e.vitaminB12), optionalNumber(e.vitaminE),
                        optionalNumber(e.vitaminK), optionalNumber(e.folate), optionalNumber(e.omega3),
                        sourceLabel(e.source), e.customNote ?? ""
                    ]
                    s += cols.map(csvEscape).joined(separator: ",") + "\n"
                }
            }
        }
        return s
    }

    private nonisolated static func csvEscape(_ field: String) -> String {
        if field.contains(",") || field.contains("\"") || field.contains("\n") {
            return "\"" + field.replacingOccurrences(of: "\"", with: "\"\"") + "\""
        }
        return field
    }
}
